test/validator.go

```diff
@@ -1,197 +1,273 @@
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
 )
 
-type Validator struct {
-	Index     int
-	Sentries  map[string]Nodes
-	Signers   Signers
-	tl        Logger
-	Home      string
-	PubKeys   map[string]cometcrypto.PubKey
-	Threshold uint8
-}
+const (
+	testChain        = "gaia" // ghcr.io/strangelove-ventures/heighliner/gaia
+	testChainVersion = "v10.0.2"
 
-func NewHorcruxValidator(
-	tl Logger,
-	pool *dockertest.Pool,
-	networkID string,
-	home string,
-	index int,
-	numSigners int,
-	threshold uint8,
-	chains ...*ChainType,
-) (*Validator, error) {
-	chainIDs := make([]string, 0, len(chains))
-	sentries := make(map[string]Nodes)
-	for _, chain := range chains {
-		sentries[chain.ChainID] = MakeNodes(
-			index,
-			chain.NumSentries,
-			home,
-			chain.ChainID,
-			chain,
-			pool,
-			networkID,
-			tl,
-		)
-		chainIDs = append(chainIDs, chain.ChainID)
-	}
-	testValidator := &Validator{
-		Index:     index,
-		Sentries:  sentries,
-		Signers:   MakeSigners(index, numSigners, home, pool, networkID, tl),
-		tl:        tl,
-		Home:      home,
-		Threshold: threshold,
-		PubKeys:   make(map[string]cometcrypto.PubKey),
-	}
-	if err := testValidator.genPrivKeyAndShards(nil, chainIDs...); err != nil {
-		return nil, err
-	}
-	return testValidator, nil
+	signerPort       = "2222"
+	signerPortDocker = signerPort + "/tcp"
+
+	signerImage        = "horcrux-test"
+	binary             = "horcrux"
+	signerImageUidGid  = "2345:2345"
+	signerImageHomeDir = "/home/horcrux"
+)
+
+// chainWrapper holds the initial configuration for a chain to start from genesis.
+type chainWrapper struct {
+	chain           **cosmos.CosmosChain
+	totalValidators int // total number of validators on chain at genesis
+	totalSentries   int // number of additional sentry nodes
+	modifyGenesis   func(cc ibc.ChainConfig, b []byte) ([]byte, error)
+	preGenesis      func(ibc.ChainConfig) error
 }
 
-func NewHorcruxValidatorWithPrivValKey(
-	tl Logger,
-	pool *dockertest.Pool,
-	networkID string,
-	home string,
-	index int,
-	numSentries int,
-	numSigners int,
-	threshold uint8,
-	chainType *ChainType,
-	privValKey privval.FilePVKey,
-) (*Validator, error) {
-	chainID := chainType.ChainID
-	sentries := make(map[string]Nodes)
-	sentries[chainID] = MakeNodes(index, numSentries, home, chainID, chainType, pool, networkID, tl)
-	testValidator := &Validator{
-		Index:     index,
-		Sentries:  sentries,
-		Signers:   MakeSigners(index, numSigners, home, pool, networkID, tl),
-		tl:        tl,
-		Home:      home,
-		Threshold: threshold,
-		PubKeys:   make(map[string]cometcrypto.PubKey),
+// startChains starts the given chains locally within docker composed of containers.
+func startChains(
+	ctx context.Context,
+	t *testing.T,
+	logger *zap.Logger,
+	client *client.Client,
+	network string,
+	chains ...chainWrapper,
```

Comment:
Changing this to `chains ...*chainWrapper` may help clean up some of the dereferncing.